### PR TITLE
Return checkers

### PR DESCRIPTION
**Pull request by baloghadamsoftware**

### CODE REVIEW

File: include/clang/StaticAnalyzer/Checkers/Checkers.td [hunk 1]
```diff
@@ -770,3 +793,20 @@ def UnixAPIPortabilityChecker : Checker<"UnixAPI">,
...
+
+//===----------------------------------------------------------------------===/
+// Statistics collectors.
+//===----------------------------------------------------------------------===/
+
+
+let ParentPackage = StatisticsCollector in {
+
+def ReturnValueCheckStatisticsCollector : Checker<"ReturnValueCheck">,
+    HelpText<"Collectes statistics about the return value checks">,
```
Comment (dkrupp) [inline include/clang/StaticAnalyzer/Checkers/Checkers.td]: typo: Collectes -> Collects

File: lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
+//==- ReturnValueCheckStatisticsCollector.cpp --------------------*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (dkrupp) [inline lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp]: Could you please add a comment describing what statistics this checker is collecting, what is the intended use, what are the input config variables.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
+//===-- SpecialReturnValueStatisticsCollector.cpp -----------------*- C++ -*--//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (dkrupp) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: please add description to this checker too

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
+//==- SpecialReturnValueChecker.cpp ------------------------------*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (dkrupp) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: please describe the configuration dependencies of this checker: the statistics file path and how that file is used

File: lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
+//==- UncheckedReturnValueChecker.cpp ----------------------------*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (dkrupp) [inline lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp]: please describe what this checker should be verifying

File: lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  for (Stmt *Child : CE->children()) {
+    if (Child) {
+      Visit(Child);
+    }
+  }
+}
+
+void ReturnValueCheckVisitor::VisitCompoundStmt(CompoundStmt *S) {
+  for (Stmt *Child : S->children()) {
+    if (Child) {
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp]: Maybe we could hoist null checks to VisitStmt and use dyn_cast_or_null bellow.

File: lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  }
+}
+
+namespace {
+class ReturnValueCheckStatisticsCollector : public Checker<check::ASTCodeBody> {
+public:
+  std::unique_ptr<BugType> UncheckedReturnValueBugType;
+
+  ReturnValueCheckStatisticsCollector() {
+    UncheckedReturnValueBugType.reset(new BugType(
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp]: Usually the name of the check is not correct in the plist, when we init the BugType in the ctor. Maybe we should use the mutable BugType + Lazy Init pattern.

File: lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        getCheckName(), "Return value unchecked", "Misuse of APIs"));
+    UncheckedReturnValueBugType->setSuppressOnSink(true);
+  }
+
+  void checkASTCodeBody(const Decl *D, AnalysisManager &mgr,
+                        BugReporter &BR) const {
+    ReturnValueCheckVisitor visitor(BR, mgr.getAnalysisDeclContext(D),
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp]: Argument and variable names should begin with uppercase letters.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    io.mapRequired("name", s.Name);
+    io.mapRequired("relation", s.Rel);
+    io.mapRequired("value", s.Val);
+  }
+};
+
+}
+}
+
+static std::map<std::string, FuncRetValToCheck> FuncRetValsToCheck;
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: Use llvm StringMap here.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    io.enumCase(value, "GE", BO_GE);
+  }
+};
+
+template <> struct ScalarTraits<llvm::APSInt> {
+  static void output(const llvm::APSInt &val, void*, llvm::raw_ostream &out) {
+    val.print(out, val.isSigned());
+  }
+  static StringRef input(StringRef scalar, void*, llvm::APSInt &value) {
+    value = llvm::APSInt(scalar);
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: Lower case variable names.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    static_cast<SVal>(loc::ConcreteInt(RefInt));
+  const auto Relation =
+    SVB.evalBinOp(State, ToCheck->second.Rel, RetVal, RefVal,
+                  SVB.getConditionType()).getAs<DefinedOrUnknownSVal>();
+  if (!Relation)
+    return;
+
+  ProgramStateRef StateError, StateNormal;
+  std::tie(StateError, StateNormal) = State->assume(*Relation);
+  if (StateError) {
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: Are we sure that we want to split the exploded graph? Will the users understand the performance implications of adding "annotations"?

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+void SpecialReturnValueChecker::checkPostCall (const CallEvent &Call,
+                                               CheckerContext &C) const {
+  if (FuncRetValsToCheck.empty())
+    return;
+
+  const auto *FD = dyn_cast_or_null<FunctionDecl>(Call.getDecl());
+  if (!FD)
+    return;
+
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: Maybe some sanity check if the return type can be converted to an integer type at all before calling expensive methods like getQualifiedNameAsString?

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+  }
+
+  void Profile(llvm::FoldingSetNodeID &ID) const {
+    ID.AddPointer(Call);
+    ID.AddBoolean(checkedForNegative);
+    ID.AddBoolean(checkedForNull);
+  }
+};
+
+std::map<const CallExpr *, std::pair<bool, bool>> CheckedCalls;
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: llvm DenseMap

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+  void checkDeadSymbols(SymbolReaper &SR, CheckerContext &C) const;
+  void checkEndOfTranslationUnit(const TranslationUnitDecl *TU,
+                                 AnalysisManager &Mgr, BugReporter &BR) const;
+};
+} // namespace
+
+REGISTER_MAP_WITH_PROGRAMSTATE(SpecialReturnValueMap, SymbolRef,
+                               SpecialReturnValue)
+
+SpecialReturnValueStatisticsCollector::SpecialReturnValueStatisticsCollector() {
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: Use lazy init.

File: lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace {
+class ReturnValueCheckStatisticsCollector : public Checker<check::ASTCodeBody> {
+public:
+  std::unique_ptr<BugType> UncheckedReturnValueBugType;
+
+  ReturnValueCheckStatisticsCollector() {
+    UncheckedReturnValueBugType.reset(new BugType(
+        getCheckName(), "Return value unchecked", "Misuse of APIs"));
+    UncheckedReturnValueBugType->setSuppressOnSink(true);
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/ReturnValueCheckStatisticsCollector.cpp]: Is supress on sink justified?

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+    return;
+
+  if (Func->getReturnType()->isVoidType())
+    return;
+
+  const auto RetSym = Call.getReturnValue().getAsSymbol();
+  if (!RetSym)
+    return;
+
+  const auto *Orig = dyn_cast_or_null<CallExpr>(Call.getOriginExpr());
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: I would move this check before getting the return value symbol.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+void SpecialReturnValueStatisticsCollector::checkDeadSymbols(
+    SymbolReaper &SR, CheckerContext &C) const {
+  auto State = C.getState();
+
+  auto SymbolMap = State->get<SpecialReturnValueMap>();
+  for (const auto Sym : SymbolMap) {
+    if (!SR.isLive(Sym.first)) {
+      CheckedCalls[Sym.second.getCall()];
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: Multiple lookups in the map. I think one could be eliminated.

Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: Are we sure that the bools are initialized to false?

File: lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+  if (!std::binary_search(APIFuncsReturningError->begin(),
+                          APIFuncsReturningError->end(), FullName))
+    return;
+
+  // Issue a warning.
+  SmallString<256> buf;
+  llvm::raw_svector_ostream os(buf);
+  os << "Return value is not checked in call to '" << *FD << '\'';
+
+  const char *bugType = "Unchecked return value";
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp]: Maybe it should be a global static StringRef instead.

Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp]: Or just the local variable could be eliminated.

File: lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+  BR.EmitBasicReport(AC->getDecl(), CN, bugType, "API", os.str(), CELoc);
+}
+
+namespace {
+class UncheckedReturnValueChecker : public Checker<check::ASTCodeBody> {
+public:
+  std::unique_ptr<BugType> UncheckedReturnValueBugType;
+
+  UncheckedReturnValueChecker() {
+    UncheckedReturnValueBugType.reset(
```
Comment (Xazax-hun) [inline lib/StaticAnalyzer/Checkers/UncheckedReturnValueChecker.cpp]: Lazy init.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp (hunk 11 above)
Comment (baloghadamsoftware) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: If the value in the code is checked as expected then there is no extra split at all, just the split is moved up in the graph. If it is not checked we expect one of the branches to sink sooner or later with an error.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp (hunk 17 above)
Comment (baloghadamsoftware) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp]: Multiple lookups are intentional. The first lookup just ensures that there is a map entry for the call. If this entry existed earlier, this statement ha no effect, the value does not change. If it did not exist a new is constructed using the default constructor of the value type, which is false for bool.

File: lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp (hunk 10 above)
Comment (baloghadamsoftware) [inline lib/StaticAnalyzer/Checkers/SpecialReturnValueChecker.cpp]: Copied from the comment header in the API.

File: include/clang/StaticAnalyzer/Checkers/LoadMetadata.h [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+  llvm::sys::path::append(DataFilePath, DataFileName);
+  // Try to open the data file.
+  auto DataFile = llvm::MemoryBuffer::getFile(DataFilePath);
+  if (DataFile.getError() == std::errc::no_such_file_or_directory)
+    return YamlReadError::FileNotFound;
+  if (!DataFile)
+    return YamlReadError::ReadFail;
+
+  // Check the version of the metadata file.
+  // The version comment should match "# [A-Za-z]+ metadata format ([\d\.]+)"
```
Comment (baloghadamsoftware) [inline include/clang/StaticAnalyzer/Checkers/LoadMetadata.h]: Not sure but I inherited this file from the Sony guy.

File: tools/statistics/gen_yaml_for_return_value_checks.sh [hunk 21]
```diff
@@ -0,0 +1,41 @@
+#!/bin/sh
+
```
Comment (dkrupp) [inline tools/statistics/gen_yaml_for_return_value_checks.sh]: This script in python would be much easier to maintain. we should rewrite it...

File: tools/statistics/gen_yaml_for_return_value_checks.sh [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+TEMP=`getopt -o t: --long threshold -n gen_yaml_for_return_value_checks.sh -- "$@"`
```
Comment (dkrupp) [inline tools/statistics/gen_yaml_for_return_value_checks.sh]: how would this work with running multiple clang instances in parallel? Should we collect the ouput of clang  instances in to an intermediary file and then, as a post processing step echo this file into this shell script which would calculate the percentages based on that?

File: include/clang/StaticAnalyzer/Checkers/LoadMetadata.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (SeanEveson) [reply include/clang/StaticAnalyzer/Checkers/LoadMetadata.h]: The reason for doing this was that (AFAIK) with the LLVM YAML library you have to know the structure upfront to be able to load it.

So if the version was in the YAML and you extended the format in a later version, then you wouldn't be able to read the version number (making it redundant).
The data for different checkers is in different files for this same reason.

**Pull request closed without merging**